src/main/java/com/google/sps/servlets/FreeTimeServlet.java

```diff
@@ -0,0 +1,211 @@
+
+    PlanMailResponse planMailResponse =
+        new PlanMailResponse(wordCount, averageReadingSpeed, minutesToRead, potentialTimes);
+    // Convert event list to JSON and print to response
+    JsonUtility.sendJson(response, planMailResponse);
+  }
+
+  private List<DateInterval> getPotentialTimes(FreeTimeUtility freeTimeUtility, long timeNeeded) {
+    List<DateInterval> workFreeInterval = freeTimeUtility.getWorkFreeInterval();
+    List<DateInterval> potentialEventTimes = new ArrayList<>();
+    long remainingTime = timeNeeded;
+    for (DateInterval interval : workFreeInterval) {
+      Date potentialEnd = new Date(interval.getStart().getTime() + remainingTime);
+      if (interval.getEnd().before(potentialEnd)) {
+        remainingTime -= (interval.getEnd().getTime() - interval.getStart().getTime());
+        potentialEventTimes.add(interval);
+      } else {
+        potentialEventTimes.add(new DateInterval(interval.getStart(), potentialEnd));
+        break;
+      }
+    }
+    return potentialEventTimes;
+  }
+
+  private int getWordCount(Credential googleCredential) {
+    GmailClient gmailClient = gmailClientFactory.getGmailClient(googleCredential);
+    GmailClient.MessageFormat messageFormat = GmailClient.MessageFormat.FULL;
+    int numberDays = 7;
+    int wordCount = 0;
+    List<Message> unreadMessages = new ArrayList<>();
+    try {
+      unreadMessages = gmailClient.getUnreadEmailsFromNDays(messageFormat, numberDays);
+    } catch (IOException e) {
+      System.out.println(e);
+    }
+    for (Message message : unreadMessages) {
+      try {
+        wordCount += getMessageSize(message);
+      } catch (MessagingException | IOException e) {
+        System.out.println(e);
+      }
+    }
+    return wordCount;
+  }
+
+  /**
+   * Get the events in the user's calendars
+   *
+   * @param calendarClient either a mock CalendarClient or a calendarClient with a valid credential
+   * @param timeMin the minimum time to start looking for events
+   * @param timeMax the maximum time to look for events
+   * @return List of Events from all of the user's calendars
+   * @throws IOException if an issue occurs in the method
+   */
+  private List<Event> getEvents(CalendarClient calendarClient, Date timeMin, Date timeMax)
+      throws IOException {
+    List<CalendarListEntry> calendarList = calendarClient.getCalendarList();
+    List<Event> events = new ArrayList<>();
+    for (CalendarListEntry calendar : calendarList) {
+      events.addAll(calendarClient.getUpcomingEvents(calendar, timeMin, timeMax));
+    }
+    return events;
+  }
+
+  private int getMessageSize(Message message) throws MessagingException, IOException {
+    List<MessagePart> messageParts = message.getPayload().getParts();
+    List<MessagePart> messageBody =
+        messageParts.stream()
+            .filter((messagePart) -> messagePart.getMimeType().equals("text/plain"))
+            .collect(Collectors.toList());
+    int size = 0;
```

Comment:
I think you could do everything in this method with a single stream operation. Take a look at `Stream.flatMap` and `Stream.reduce` for inspiration.